### PR TITLE
Add user pointer to RzDebug Plugins and refactor GDB to use it

### PR DESCRIPTION
**Pull request by thestr4ng3r**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

* Renamed the current debug plugin in `RzDebug` from `h` to `cur` (like in `RzAsm` for example)
* Added a plugin context for debug plugins, naming/structure implemented as in #808
* Refactored remote gdb to use a local context instead of global vars

**Test plan**

This should be a pure refactoring without noticeable behavior changes. To test remote gdb, on one machine:
`gdbserver :1337 ./executable`
And on another (or the same):
`rz gdb://<hostname>:1337`
Then debug around. The behavior should be the same before and after these changes.

### CODE REVIEW

**Review by wargio — APPROVED** [review]

File: librz/debug/p/debug_gdb.c [hunk 1]
```diff
@@ -15,86 +15,109 @@ typedef struct {
...
 #define SUPPORTED   1
 
-static RzIOGdb **origriogdb = NULL;
-static libgdbr_t *desc = NULL;
-static ut8 *reg_buf = NULL;
-static int buf_size = 0;
-static int support_sw_bp = UNKNOWN;
-static int support_hw_bp = UNKNOWN;
+typedef struct rz_debug_gdb_ctx_t {
+	RzIOGdb **origriogdb;
```
Comment (XVilka) [inline librz/debug/p/debug_gdb.c]: Maybe rename this variable too?

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
It fails MSVC build:
```
FAILED: librz/io/librz_io.a.p/p_io_debug.c.obj 
"cl" "-Ilibrz/io\librz_io.a.p" "-I." "-I.." "-I..\librz\include" "-I..\shlr\bochs\include" "-I..\shlr\gdb\include" "-I..\shlr\gdb\include\gdbclient" "-I..\shlr\gdb\include\gdbserver" "-I..\shlr\winkd" "-I..\shlr\qnx\include" "-I..\subprojects\libzip-1.7.3\lib" "-Isubprojects\libzip-1.7.3" "-I..\subprojects\libzip-1.7.3" "-Isubprojects\zlib-1.2.11" "-I..\subprojects\zlib-1.2.11" "-I..\shlr\ar" "-Isubprojects\sdb" "-I..\subprojects\sdb" "-Isubprojects\sdb\src" "-I..\subprojects\sdb\src" "-I..\shlr\spp" "-I..\shlr\w32dbg_wrap\include" "/MT" "/nologo" "/showIncludes" "/W2" "/O2" "/Gw" "-DRZ_PLUGIN_INCORE=1" "/Fdlibrz/io\librz_io.a.p\p_io_debug.c.pdb" /Folibrz/io/librz_io.a.p/p_io_debug.c.obj "/c" ../librz/io/p/io_debug.c
../librz/io/p/io_debug.c(532): error C2039: 'user': is not a member of 'rz_debug_t'
..\librz\include\rz_debug.h(240): note: see declaration of 'rz_debug_t'
../librz/io/p/io_debug.c(561): error C2039: 'user': is not a member of 'rz_debug_t'
..\librz\include\rz_debug.h(240): note: see declaration of 'rz_debug_t'
[438/1620] Compiling C object librz/io/librz_io.a.p/p_io_bochs.c.obj
[439/1620] Compiling C object librz/io/librz_io.a.p/p_io_default.c.obj
ninja: build stopped: subcommand failed.
```

**Review by yossizap — APPROVED** [review]
LGTM tested in Ubuntu

File: librz/debug/p/debug_gdb.c (hunk 1 above)
Comment (thestr4ng3r) [reply librz/debug/p/debug_gdb.c]: Rename to what exactly?

**Comment by XVilka** [comment]
Broken debug tests on Windows:
```
[XX] C:\projects\rizin\test\db\archos\windows-x64\dbg_rebase function rebase
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -d -Qc '?v main-`e bin.baddr`
doc
?v main-`e bin.baddr`
' bins/pe/hello_world/hello_world_od64bit.exe
-- stdout
--- expected
+++ actual
@@ -1,2 +1,0 @@
-0x1000
-0x1000
-- stderr
Spawned new process with pid 3188, tid = 3300
= attach 3188 3300
bin.baddr 0x7ff7abe70000
Using 0x7ff7abe70000
-- exit status: -1073741819
[XX] C:\projects\rizin\test\db\archos\windows-x64\dbg_rebase ref rebase
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -d -Qc 'aa
?v `axt main~entry0[1]`-`e bin.baddr`
doc
?v `axt main~entry0[1]`-`e bin.baddr`
' bins/pe/hello_world/hello_world_od64bit.exe
-- stdout
--- expected
+++ actual
@@ -1,2 +1,0 @@
-0x1203
-0x1203
-- stderr
Spawned new process with pid 5220, tid = 2536
= attach 5220 2536
bin.baddr 0x7ff7abe70000
Using 0x7ff7abe70000
-- exit status: -1073741819
[XX] C:\projects\rizin\test\db\archos\windows-x64\dbg_rebase bp rebase
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -d -Qc 'db main
?v main-`db~main[0]`
doc
?v main-`db~main[0]`
' bins/pe/hello_world/hello_world_od64bit.exe
-- stdout
--- expected
+++ actual
@@ -1,2 +1,0 @@
-0x0
-0x0
-- stderr
Spawned new process with pid 5500, tid = 2288
= attach 5500 2288
bin.baddr 0x7ff7abe70000
Using 0x7ff7abe70000
-- exit status: -1073741819
[XX] C:\projects\rizin\test\db\archos\windows-x64\dbg_rebase var rebase
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -d -Qc 'aa
afv @ entry0
doc
afv @ entry0
' bins/pe/hello_world/hello_world_od64bit.exe
-- stdout
--- expected
+++ actual
@@ -1,6 +1,0 @@
-var int64_t var_20h @ rsp+0x48
-var int64_t var_8h @ rsp+0x68
-var int64_t var_10h @ rsp+0x70
-var int64_t var_20h @ rsp+0x48
-var int64_t var_8h @ rsp+0x68
-var int64_t var_10h @ rsp+0x70
-- stderr
Spawned new process with pid 4904, tid = 4868
= attach 4904 4868
bin.baddr 0x7ff7abe70000
Using 0x7ff7abe70000
-- exit status: -1073741819
[**]                 C:\projects\rizin\test\db\asm\8051    13544 OK       921 BR        4 XX       18 FX
[XX] C:\projects\rizin\test\db\archos\windows-x64\dbg_rebase flag rebase
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -d -Qc 'fs test
f testflag @ main+10
?v `f~testflag[0]`-`e bin.baddr`
doc
?v `f~testflag[0]`-`e bin.baddr`
' bins/pe/hello_world/hello_world_od64bit.exe
-- stdout
--- expected
+++ actual
@@ -1,2 +1,0 @@
-0x100a
-0x100a
-- stderr
Spawned new process with pid 3952, tid = 1928
= attach 3952 1928
bin.baddr 0x7ff7abe70000
Using 0x7ff7abe70000
-- exit status: -1073741819
[**] C:\projects\rizin\test\db\archos\windows-x64\dbg_rebase    13544 OK       921 BR        5 XX       18 FX
[**]             C:\projects\rizin\test\db\analysis\z80    13545 OK       921 BR        5 XX       18 FX
[**]          C:\projects\rizin\test\db\analysis\xtensa    13547 OK       921 BR        5 XX       18 FX
[XX] C:\projects\rizin\test\db\archos\windows-x64\dbg_oo ood and doc
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '(reopen; ood > $_; doc > $_)
4 .(reopen)
' bins\pe\hello_world_not_stripped.exe
-- stderr
Spawned new process with pid 5824, tid = 2236
= attach 5824 2236
-- exit status: -1073741819
[**] C:\projects\rizin\test\db\archos\windows-x64\dbg_oo    13547 OK       921 BR        6 XX       18 FX
[XX] C:\projects\rizin\test\db\archos\windows-x64\dbg_dts continue backwards
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e dbg.create_new_console=false
ood
db main
3dc
dr rflags=0x200 > nul
dr rflags,rax
dts+
7ds
dr rflags,rax
dcb
dr rflags,rax
doc
' bins/pe/ioli/w32/crackme0x00.exe
-- stdout
--- expected
+++ actual
@@ -1,6 +1,0 @@
-0x00000200
-0x00000001
-0x00000216
-0x0000001e
-0x00000200
-0x00000001
-- stderr
Spawned new process with pid 5644, tid = 1336
= attach 5644 1336
-- exit status: -1073741819
[**] C:\projects\rizin\test\db\archos\windows-x64\dbg_dts    13548 OK       921 BR        7 XX       18 FX
[**]          C:\projects\rizin\te
```

**Review by XVilka — CHANGES_REQUESTED** [review]
.

File: librz/include/rz_debug.h [hunk 2]
```diff
@@ -358,11 +359,12 @@ typedef struct rz_debug_plugin_t {
 	const char *license;
 	const char *author;
 	const char *version;
-	//const char **archs; // MUST BE DEPRECATED!!!!
 	ut32 bits;
 	const char *arch;
 	int canstep;
 	int keepio;
+	int (*init)(RzDebug *dbg, void **user);
```
Comment (ret2libc) [inline librz/include/rz_debug.h]: Maybe add typedef func type?

**Review by ret2libc — APPROVED** [review]

**Comment by thestr4ng3r** [comment]
This might be because of the changes in `rz_debug_use()`, I'll check it.

File: librz/include/rz_debug.h (hunk 2 above)
Comment (thestr4ng3r) [reply librz/include/rz_debug.h]: What do you mean exactly? All the other plugin callbacks are defined in-place too so I am not sure why we would want to use a typedefd one here.

Comment (wargio) [reply librz/include/rz_debug.h]: please use bool and not int

**Comment by thestr4ng3r** [comment]
Should hopefully be fixed now, the windows debug plugin was depending on the io plugin's init to have been run first.

File: librz/include/rz_debug.h (hunk 2 above)
Comment (thestr4ng3r) [reply librz/include/rz_debug.h]: changed to bool

**Review by XVilka — APPROVED** [review]

**Comment by wargio** [comment]
```
[XX] db/io/http r2 http://example.org
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'r
b 64
ps
' http://example.org
-- stdout
--- expected
+++ actual
@@ -1,6 +1,0 @@
-1256
-<!doctype html>
-<html>
-<head>
-    <title>Example Domain</title>
-

```
WTH

**Comment by thestr4ng3r** [comment]
Yes, our tests depend on the availability of http://example.org.

**Review by ret2libc — APPROVED** [review]

**Pull request merged**